src/cuda/cuda_sampling.cu

```diff
@@ -498,13 +509,129 @@ __global__ void GetTopKKernel(int* indices_out, float* scores_in, float* scores_
+template <int kBlockSize>
+__global__ void GetTopKKernelDistributed(int* indices_out, float* scores_in, float* scores_out, int batch_size, 
+                                        int vocab_size, int k, float temperature, int top_k_shards, 
+                                        int* top_k_distributed_lock, int* distributed_indices_out, float* distributed_scores_out) {
+  int batch = blockIdx.x;
+  int tid = threadIdx.x;
+  TopK_2 partial;
+
+  float const MAX_T_VAL = FLT_MAX;
+
+  int* distributed_indices_out_curr = distributed_indices_out + blockIdx.z * k;
+  float* distributed_scores_out_curr = distributed_scores_out + blockIdx.z * k;
+
+  int start_i = blockIdx.z * blockDim.x + tid;  
+ 
+  for (int ite = 0; ite < k; ite++) {
+    partial.init();
+
+    for (auto elemId = start_i; elemId < vocab_size; elemId += kBlockSize* gridDim.z) {
+      float elem = scores_in[elemId + batch * vocab_size];
+      partial.insert(elem, elemId);
+    }
+    // reduce in thread block
+    typedef cub::BlockReduce<TopK_2, kBlockSize> BlockReduce;
+    __shared__ typename BlockReduce::TempStorage temp_storage;
+    TopK_2 top_k_sequence = BlockReduce(temp_storage).Reduce(partial, reduce_topk_op_2);
+
+    if (tid == 0) {
+      // No temperature scaling here - we will do it after the final reduction
+      distributed_scores_out_curr[ite + batch * k] = top_k_sequence.u;
+
+      distributed_indices_out_curr[ite + batch * k] = top_k_sequence.p;
+
+      // set the max value to -MAX_T_VAL so that the value doesn't get picked again
+      scores_in[batch * vocab_size + top_k_sequence.p] = -MAX_T_VAL;
+      __threadfence_block();
+    }
+
+    __syncthreads();
+  }
+
+  // All TBs flush their data and it should be visible to every other TB
+  __threadfence();   
+  __syncthreads();
+
+  // Signal that each threadblock has done its work using elected thread
+  if (threadIdx.x == 0)
+  {
+    atomicAdd(top_k_distributed_lock, 1);
+  }
+
+  // The reduction threadblock
+  if (blockIdx.z == 0) {
+    // Elected thread spins while waiting for other TBs to complete their work
+    if (threadIdx.x == 0)
+    {
+      int count_of_completed_TBs = 0;
+
+      asm volatile("ld.volatile.global.s32 %0, [%1];" : "=r"(count_of_completed_TBs) : "l"(top_k_distributed_lock));
+      while (count_of_completed_TBs < top_k_shards)
+      {
+          asm volatile("ld.volatile.global.s32 %0, [%1];" : "=r"(count_of_completed_TBs) : "l"(top_k_distributed_lock));
+      }
+
+      // Reset the lock for next kernel call
+      atomicExch(top_k_distributed_lock, 0);
+    }
+
+    // TODO: Is there enough shared memory to bring the top_k shards to shared memory on all GPUs?
+    TopK_2 reduce_partial; 
+    for (int ite = 0; ite < k; ite++) {
+      reduce_partial.init();
+  
+      for (int i = threadIdx.x; i < top_k_shards*k; i += blockDim.x) {
+        float score = distributed_scores_out[i];
+        reduce_partial.insert(score, i);
```

Comment:
i is the buffer index instead of vocab index, this will cause tie-breaking error (so the sort result is no longer stable).